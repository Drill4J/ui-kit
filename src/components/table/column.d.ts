import { ColumnProps } from './table-types';

export declare const Column: {
  (props: ColumnProps): null;
  displayName: string;
  __docgenInfo: {
    description: string;
    displayName: string;
    props: {
      name: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      Cell: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      HeaderCell: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      label: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      align: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
    };
  };
};
