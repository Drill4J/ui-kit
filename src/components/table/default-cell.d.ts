export declare const DefaultCell: {
    ({ value }: {
        value: unknown;
    }): string | null;
    displayName: string;
    __docgenInfo: {
        description: string;
        displayName: string;
        props: {
            value: {
                defaultValue: null;
                description: string;
                name: string;
                required: boolean;
                type: {
                    name: string;
                };
            };
        };
    };
};
