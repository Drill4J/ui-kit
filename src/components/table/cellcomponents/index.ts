import { Name } from './name';
import { Link } from './link';
import { AdditionalInfo } from './additional-info';

export const CellComponents = {
  Link,
  Name,
  AdditionalInfo,
} as any;
