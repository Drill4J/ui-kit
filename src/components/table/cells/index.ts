export { DurationCell as Duration } from './duration-cell';
export { CompoundCell as Compound } from './compound-cell';
export { CoverageCell as Coverage } from './coverage-cell';
export { ClickableCell as Clickable } from './clickable-cell';
export { TestStatusCell as TestStatus } from './test-status-cell';
