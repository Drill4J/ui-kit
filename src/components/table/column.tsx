import { ColumnProps } from './table-types';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const Column = (props: ColumnProps) => null;
