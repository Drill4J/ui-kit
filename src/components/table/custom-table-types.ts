import { Column } from 'react-table';

export type CustomColumn = Column &
{ textAlign?: 'center' | 'right' | 'left';
  width?: string;
  notSortable?: boolean;
  disableEllipsis?: boolean;
  filterable?: boolean;
  isCustomCell?: boolean
};
