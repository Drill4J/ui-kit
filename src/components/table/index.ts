export { Table } from './table';
export { Column } from './column';
export { SelectableTable } from './selectable-table';
export { ExpandableTable } from './expandable-table';
