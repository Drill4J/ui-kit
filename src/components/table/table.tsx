import * as React from 'react';
import { BEM } from '@redneckz/react-bem-helper';

import { TableRow } from './table__row';
import { TableHeader } from './table__header';

import styles from './table.module.scss';

interface Props {
  className?: string;
  data?: Array<{ [key: string]: unknown }>;
  children: any;
  idKey?: string;
  footer?: React.ReactNode;
  sort?: object;
  onSort?: (sortField: string) => void;
  columnsSize?: 'wide' | 'medium';
  expandedRows?: string[];
  expandedColumns?: any[];
  secondLevelExpand?: any[];
  expandedContentKey?: string;
  withoutHeader?: boolean;
  selectedRows?: string[];
}

const table = BEM(styles);

export const Table = table(
  ({
    className,
    data = [],
    children,
    idKey = '',
    footer,
    expandedRows = [],
    expandedColumns,
    expandedContentKey,
    secondLevelExpand,
    withoutHeader,
    selectedRows = [],
  }: Props) => {
    const columns = React.Children.map(children, (column) => column && column.props);
    const expandedColumnsComponents = React.Children.map(
      expandedColumns,
      (column) => column && column.props,
    );
    const expandedColumnsSecondLevel = React.Children.map(
      secondLevelExpand,
      (column) => column && column.props,
    );

    return (
      <table className={className}>
        {!withoutHeader && <TableHeader columns={columns} />}
        <tbody>
          {data.map((item, index) => (
            <TableRow
              key={idKey ? String(item[idKey]) : index}
              item={item}
              columns={columns}
              index={index}
              expandedColumns={expandedColumnsComponents}
              color={getRowColor({ expandedRows, selectedRows, itemId: String(item[idKey]) })}
              expandedContentKey={expandedContentKey}
              expandedRows={expandedRows}
              secondLevelExpand={expandedColumnsSecondLevel}
            />
          ))}
        </tbody>
        {footer}
      </table>
    );
  },
);

// eslint-disable-next-line consistent-return
function getRowColor({
  expandedRows,
  selectedRows,
  itemId,
}: {
  expandedRows: string[];
  selectedRows: string[];
  itemId: string;
}) {
  if (expandedRows.includes(itemId)) {
    return 'blue';
  }
  if (selectedRows.includes(itemId)) {
    return 'yellow';
  }
}
