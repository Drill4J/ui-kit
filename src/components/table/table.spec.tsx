/*
 * Copyright 2020 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import {
  fireEvent, render, screen,
} from '@testing-library/react';
import { Table } from './table';
import { TableActionsProvider } from './table-actions';
import { getAllByDataTest, getByDataTest, queryByDataTest } from '../../../test-utils';

const createData = (length: number) => Array.from({ length }, (_, i) => ({
  buildVersion: `0.${i}.0`,
  detectedAt: 1625224274896,
  summary: null,
}));

const props = {
  isDefaulToggleSortBy: true,
  columns: [
    {
      Header: 'Name',
      accessor: 'buildVersion',
      Cell: ({ value: buildVersion }: any) => (
        <div tw="grid gap-x-2 h-12 items-center font-bold text-14">
          <div tw="link text-ellipsis">{buildVersion}</div>
        </div>
      ),
      textAlign: 'left',
      width: '30%',
    },
    {
      Header: 'Added',
      accessor: 'detectedAt',
      textAlign: 'left',
      width: '20%',
    },
    {
      Header: 'Total methods',
      accessor: 'summary.total',
      width: '10%',
    },
    {
      Header: 'New',
      accessor: 'summary.new',
      width: '10%',
    },
    {
      Header: 'Modified',
      accessor: 'summary.modified',
      width: '10%',
    },
    {
      Header: 'Unaffected',
      accessor: 'summary.unaffected',
      width: '10%',
    },
    {
      Header: 'Deleted',
      accessor: 'summary.deleted',
      width: '10%',
    },
  ],
  data: createData(26),
};

describe('Table', () => {
  it('should render table with search if props "withSearch" is equal true', () => {
    render(<TableActionsProvider><Table {...props} withSearch placeholder="table placeholder" /></TableActionsProvider>);
    expect(screen.getByPlaceholderText('table placeholder')).toBeInTheDocument();
  });

  it('should render table without search if props "withSearch" is equal false', () => {
    render(<TableActionsProvider><Table {...props} withSearch={false} placeholder="table placeholder" /></TableActionsProvider>);
    expect(screen.queryByPlaceholderText('table placeholder')).toBeNull();
  });

  it('should render table with stub if it passed via props', () => {
    const { rerender } = render(<TableActionsProvider><Table {...props} stub={<div>stub</div>} /></TableActionsProvider>);
    expect(screen.getByText('stub')).toBeInTheDocument();

    rerender(<TableActionsProvider><Table {...props} /></TableActionsProvider>);
    expect(screen.queryByText('stub')).toBeNull();
  });

  it('should not render pagination if rows less than 25', () => {
    const { container } = render(<TableActionsProvider><Table {...props} data={createData(24)} /></TableActionsProvider>);
    expect(queryByDataTest(container, 'table:displaying-results-count')).toBeNull();
    expect(queryByDataTest(container, 'table:rows:count')).toBeNull();
    expect(queryByDataTest(container, 'table-pagination:pagination-arrow-left')).toBeNull();
    expect(queryByDataTest(container, 'table-pagination:pagination-arrow-right')).toBeNull();
  });

  it('should render 25 rows by default data.length more than 25', () => {
    render(<TableActionsProvider><Table {...props} data={createData(26)} /></TableActionsProvider>);
    expect(screen.getAllByText('1625224274896').length).toBe(25);
  });

  it('should render 50 rows if select it in select rows dropdown', () => {
    const { container } = render(<TableActionsProvider><Table {...props} data={createData(51)} /></TableActionsProvider>);
    fireEvent.click(getByDataTest(container, 'table-pagination:select-rows-dropdown'));
    fireEvent.click(getAllByDataTest(container, 'table-pagination:select-rows-dropdown:item')[1]);
    expect(screen.getAllByText('1625224274896').length).toBe(50);
  });

  it('should throw error when data is not an object and display stub', () => {
    // @ts-ignore
    render(<TableActionsProvider><Table {...props} data="" /></TableActionsProvider>);
    expect(screen.getByText('Table load failed')).toBeInTheDocument();
  });

  // it('should sort table after click to table cell', () => {
  //   const { container } = render(<TableActionsProvider><Table {...props} isDefaulToggleSortBy /></TableActionsProvider>);
  //   const headerCell = getByDataTest(container, 'table-th-buildVersion');
  //   expect(screen.getAllByText(/0./)[0].textContent).toBe('0.0.0');
  //   fireEvent.click(getByDataTest(headerCell, 'table-th-sort-arrow'));
  //   expect(screen.getAllByText(/0./)[0].textContent).toBe('0.24.0');
  // });
});
