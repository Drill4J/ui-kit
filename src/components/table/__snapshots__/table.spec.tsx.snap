// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot 1`] = `
<table
  className="table"
>
  <thead
    className="table__header"
  >
    <tr>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div
          className="panel"
          data-test=""
          onClick={[Function]}
        >
          <div
            className="sort-arrows"
          >
            <div
              className="sort-arrows__sort-arrows-wrapper"
              data-modifiers=""
            >
              <div
                className="sort-arrows__sort-arrow"
              >
                <svg
                  className="generic-icon"
                  data-modifiers=""
                  fillRule="evenodd"
                  height={10}
                  style={
                    Object {
                      "WebkitTransform": "rotate(180deg)",
                      "opacity": "100%",
                    }
                  }
                  transform="rotate(180)"
                  viewBox="0 0 7 10"
                  width={7}
                >
                  <path
                    d="M3.5 10l.01-.001h.033l.03-.006.017-.001.015-.003-.062.01a.5.5 0 0 0 .31-.145l2.95-2.95a.5.5 0 0 0-.707-.707L4 8.295 4 .5a.5.5 0 0 0-1 0l-.001 7.843L.854 6.197a.5.5 0 0 0-.708.707l2.95 2.95a.498.498 0 0 0 .311.144L3.49 10a.51.51 0 0 0 .011 0z"
                  />
                </svg>
              </div>
              <div
                className="sort-arrows__sort-arrow"
              >
                <svg
                  className="generic-icon"
                  data-modifiers=""
                  fillRule="evenodd"
                  height={10}
                  style={
                    Object {
                      "WebkitTransform": "rotate(0deg)",
                      "opacity": "100%",
                    }
                  }
                  transform="rotate(0)"
                  viewBox="0 0 7 10"
                  width={7}
                >
                  <path
                    d="M3.5 10l.01-.001h.033l.03-.006.017-.001.015-.003-.062.01a.5.5 0 0 0 .31-.145l2.95-2.95a.5.5 0 0 0-.707-.707L4 8.295 4 .5a.5.5 0 0 0-1 0l-.001 7.843L.854 6.197a.5.5 0 0 0-.708.707l2.95 2.95a.498.498 0 0 0 .311.144L3.49 10a.51.51 0 0 0 .011 0z"
                  />
                </svg>
              </div>
            </div>
          </div>
          Name
        </div>
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div
          className="panel"
          data-test=""
          onClick={[Function]}
        >
          <div
            className="sort-arrows"
          >
            <div
              className="sort-arrows__sort-arrows-wrapper"
              data-modifiers=""
            >
              <div
                className="sort-arrows__sort-arrow sort-arrows__sort-arrow--active"
              >
                <svg
                  className="generic-icon"
                  data-modifiers=""
                  fillRule="evenodd"
                  height={10}
                  style={
                    Object {
                      "WebkitTransform": "rotate(180deg)",
                      "opacity": "100%",
                    }
                  }
                  transform="rotate(180)"
                  viewBox="0 0 7 10"
                  width={7}
                >
                  <path
                    d="M3.5 10l.01-.001h.033l.03-.006.017-.001.015-.003-.062.01a.5.5 0 0 0 .31-.145l2.95-2.95a.5.5 0 0 0-.707-.707L4 8.295 4 .5a.5.5 0 0 0-1 0l-.001 7.843L.854 6.197a.5.5 0 0 0-.708.707l2.95 2.95a.498.498 0 0 0 .311.144L3.49 10a.51.51 0 0 0 .011 0z"
                  />
                </svg>
              </div>
              <div
                className="sort-arrows__sort-arrow"
              >
                <svg
                  className="generic-icon"
                  data-modifiers=""
                  fillRule="evenodd"
                  height={10}
                  style={
                    Object {
                      "WebkitTransform": "rotate(0deg)",
                      "opacity": "100%",
                    }
                  }
                  transform="rotate(0)"
                  viewBox="0 0 7 10"
                  width={7}
                >
                  <path
                    d="M3.5 10l.01-.001h.033l.03-.006.017-.001.015-.003-.062.01a.5.5 0 0 0 .31-.145l2.95-2.95a.5.5 0 0 0-.707-.707L4 8.295 4 .5a.5.5 0 0 0-1 0l-.001 7.843L.854 6.197a.5.5 0 0 0-.708.707l2.95 2.95a.498.498 0 0 0 .311.144L3.49 10a.51.51 0 0 0 .011 0z"
                  />
                </svg>
              </div>
            </div>
          </div>
          Added
        </div>
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div>
          Total Methods
        </div>
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div>
          New
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="table__row"
    >
      <td>
        <div
          onClick={[Function]}
        >
          1658090997
        </div>
      </td>
      <td>
        <span>
          1580285098638
        </span>
      </td>
      <td>
        97
      </td>
      <td>
        40
      </td>
    </tr>
  </tbody>
</table>
`;
